### PR TITLE
Portable Software Instance

### PR DESCRIPTION
**Pull request by muzikovam**
Please check if your PR fulfills the following requirements. This is simply a reminder of what we are going to look for before merging your PR. If you don't know all of this information when you create this PR, don't worry. You can edit this template as you're working on it.

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Necessary documentation (if appropriate) have been added / updated
- [x] DCO signoffs have been added to all commits, including this PR

#### PR type
What type of changes does your PR introduce to Zowe? _Put an `x` in the box that applies to this PR. If you're unsure about any of them, don't hesitate to ask._ 
- [ ] Bugfix 
- [x] Feature 
- [ ] Other... Please describe:

#### Relevant issues

Fixes  #1174

#### Changes proposed in this PR

- Added automation for Portable Software Instance

#### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

#### Does this PR do something the person installing Zowe should know about?

---
* Affected function: New way to install ZOWE*

---
* Description: Installation of ZOWE (SMP/E with latest PTFs) via Portable Software Instance including configuration workflows. *

---
* Part: _name of customizable file involved_  *

---
_multi-line description_

#### Is there a related doc issue or Pull Request? 

Doc issue/PR number: 

#### Other information

Doc https://github.com/zowe/zowe-install-packaging/blob/master/zowe_pswi_tech_prev.md

### CODE REVIEW

File: pswi/PSWI-marist.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+export TVOL=$CSIVOL
+export DVOL=$CSIVOL
+export MOUNTPATH=$ZOWE_MOUNT
+#PTF
+export CSI=$CSIHLQ
+export PTFDATASET="ZOWEAD2.ZOWE.AZWE001"
+export TARGET=$TZONE
+export DISTRIBUTION=$DZONE
+export PTF1="UO01994"
+export PTF2="UO01995"
```
Comment (jackjia-ibm) [inline pswi/PSWI-marist.sh:36]: I have a question, is this hardcoded to always build PSWI on top of Zowe v1.25.0?

Do you think it makes sense to dynamically pick up the SMPE build from previous stage?

Comment (muzikovam) [reply pswi/PSWI-marist.sh:36]: Yes, it is hardcoded there as we are now not downloading anything. With the previous stage is problem that it is not creating classic SMP/E environment. Which we are now creating here by our workflows from datasets that are already on zzow3. Same with the application of PTFs, the datasets are there.

**Comment by jackjia-ibm** [comment]
@muzikovam I'm thinking to work on some improvements for this PR. The first part is downloading the PSWI build output from zOS and uploading to Zowe artifactory. The changes are in https://github.com/zowe/zowe-install-packaging/pull/2477 but I noticed build failure and need to investigate.

The other part of the change I would like to make is using the SMPE build generated by the previous step instead of using hardcoded PTF stored on zOS. This will allow any branch to verify if the branch changes are compatible with PSWI. This is common practice for our current pipeline.

So my understanding for this PR is it doesn't matter whether it's merged or not because it always uses same inputs, so we can leave it in this branch for now. If we want to use this to build a PSWI, we can run the build from this PR.

Please let me know if this sounds ok.

**Comment by OnnoVdT** [comment]
For picking up the build output, know that the SMP/E build will always create RELFILEs & SMPMCS that have the latest code. At the moment though it does NOT tag the SMPMCS that it matches a certain PTF/APAR level. So if it is difficult for your coding to decide if it is just GA or if it has a PTF/APAR/USERMOD to pick up (do we even support PSWI with an APAR/USERMOD build?), then we can look into enhancing the RELFILE build (smpe-fmid.sh) to include PTF info, and then all you need is the RELFILEs for PSWI.

**Comment by pavelzlatnik** [comment]
> @muzikovam I'm thinking to work on some improvements for this PR. The first part is downloading the PSWI build output from zOS and uploading to Zowe artifactory. The changes are in #2477 but I noticed build failure and need to investigate.
> 
> The other part of the change I would like to make is using the SMPE build generated by the previous step instead of using hardcoded PTF stored on zOS. This will allow any branch to verify if the branch changes are compatible with PSWI. This is common practice for our current pipeline.
> 
> So my understanding for this PR is it doesn't matter whether it's merged or not because it always uses same inputs, so we can leave it in this branch for now. If we want to use this to build a PSWI, we can run the build from this PR.
> 
> Please let me know if this sounds ok.
----
Hi Jack,
To build PSWI, we need SMPE DATASETS on mainframe with applied PTFs in them. So it is not possible to use just SMPE  build PAX. It is useless for this purpose. 1th thing, SMPE pax is always the same, because changes are not applied in it (so there is no meaning to download/refresh every time the same file). Changes are done in PTFs, that is the reason why we first need to install SMPE PAX. (Always the same one). After we have SMPE environment (we are generating from SMPE PAX via workflow), over this environment we are applying PTFs (which are changing from release to release). So with every new release, we need to download official PTFs which are available on download site, and apply those PTFs into the smpe environment. After this process, we are able from that smpe environment generate PSWI. So just SMPE PAX which is generated in previous step is useless from this context. 

If this PR is not merged, we are not able to build PSWI automatically after ZOWE release is done. (Of course we have to always download new set of PTFs, and change variables export PTF1="UO01994" export PTF2="UO01995" with every new release, but the rest is already fully automated, and this part is hard to automate if we do not know what are the future numbers and links every time, second thing is that as PAX, so PTFs are really huge, and even when done download and upload manually, there were issues with space in target system, or failed FTP connection during download). 

If we would have to always generate PSWI manually, there is a lot of steps which have to be performed by someone from our team who does understand what have to be done, so it is really show-stopper for PSWI available on zowe download page.

I hope it is more clear now.
Pavel

**Comment by jackjia-ibm** [comment]
Thanks Pavel for the explanation. I understand the part FMID is always same, the difference will be PTF. What I really mean is this should pick up the PTF built from previous stage instead of using hardcoded PTF number and build on the zOS system.

Let me try to explain how current pipeline works:

- A. Build pipeline, which will run on every PR and staging, rc, master branches.
  1. upload content to zOS
  2. build pax and SMPE/PTF
  3. download build results (pax, ptf, etc) to Jenkins build server
  4. upload to Artifactory
- B. Release pipeline, which will only run few times **during release time** to promote RCs and final release.
  1. locate the artifacts in Artifactory snapshot
  2. promote (copy the artifacts from snapshot to release folder)

So here are my concerns, especially on your statement `So with every new release, we need to download official PTFs which are available on download site, and apply those PTFs into the smpe environment.`.

- This means you need official PTFs which doesn't exist before final release to build PSWI.
- If so, this step doesn't make sense to be part of build pipeline.
- In the build pipeline, you won't have official release until some artifacts are promoted,
- The PSWI build cannot be verified until final release is done. If we see some errors in the PSWI build, we don't have chance to fix it except for creating a hotfix like v1.26.1, v1.26.2, because v1.26.0 is already finalized/promoted.
- there is manual step on the fully automated pipeline to put official PTFs onto the build server before kick off this build.
- there is lack of step to promote PSWI artifact like Zowe convenience build (pax) and SMPE PTF build.

As I mentioned previously, this PR uses fixed input stored on the build server. So it doesn't matter this PR is merged or not. You can kick off a PSWI build from this PR anytime if you want this part of automation.

**Comment by pavelzlatnik** [comment]
> As I mentioned previously, this PR uses fixed input stored on the build server. So it doesn't matter this PR is merged or not. You can kick off a PSWI build from this PR anytime if you want this part of automation.

Jack I understand your concerns, and I will try to reply to them:
1)
•	_This means you need official PTFs which doesn't exist before final release to build PSWI.
•	If so, this step doesn't make sense to be part of build pipeline._

Why it does not make sense to be part of build pipeline? It is a build of ZOWE PSWI, what is new zowe artifact, and it is equivalent of smpe with applied PTFs. (To what other pipeline you would put build of ZOWE PSWI? )

2) _In the build pipeline, you won't have official release until some artifacts are promoted_

Yes, it is true, but I do not see big issue to publish PSWI for example 1 day later after SMPE release is public. Or when are PTFs ready, first could be done 1 day postpone of public publish of new convenient &smpe build, until also PSWI is generated, if is issue to publish 1 artifact day later than others.

3) _The PSWI build cannot be verified until final release is done. If we see some errors in the PSWI build, we don't have chance to fix it except for creating a hotfix like v1.26.1, v1.26.2, because v1.26.0 is already finalized/promoted._

Yes, it is true. But the build of new PSWI could be simply re-run (after issue correction), and new PSWI would be published with version 1.26.1, 1.26.2, etc. as you are stating, so I do not see any issue with it.

4) _there is manual step on the fully automated pipeline to put official PTFs onto the build server before kick off this build._

Yes, it is true. But PSWI build is manual now (it is not triggered automatically, you have to set you want to build new PSWI). So whoever would set it for run, he have to be sure why he is doing it. If proper PTFs are in place, and if in need new one, that they are downloaded&uploaded properly.
I agree that this is candidate for enhancement in some future release, to have also automatic way for PTFs handlings. 
But as I stated before, it is not much reliable with connection during download/upload, you have to be sure there is enough space in Marist system for automatic PTFs, and also it is very time consuming. Especially when you need to rerun the build if failed from various reasons.

5) _there is lack of step to promote PSWI artifact like Zowe convenience build (pax) and SMPE PTF build._

I agree, no one from my team know how to do that properly automatically in zowe machine. So if you know how to do that, you or we could add it to the pipeline.

6) _As I mentioned previously, this PR uses fixed input stored on the build server. So it doesn't matter this PR is merged or not. You can kick off a PSWI build from this PR anytime if you want this part of automation._

If we do not merge it, if anyone would delete this branch, all work will be lost. So in this case we would have to create some other repository where will be kept code for ZOWE PSWI handling. There are already issues with z/osmf versions PSWI compatibility (it is not backward compatible, so PSWI from version 1.24 cannot be loaded in z/osmf 1.23, workflows and some APIs are handled differently, etc.) And if we should also do some enhancements from release to release of zowe, we need to handle that properly in some master for every release. And I do not know if is it proper way to do open source development of ZOWE PSWI in some separate repository from other zowe artifacts build code.

**Comment by pavelzlatnik** [comment]
> For picking up the build output, know that the SMP/E build will always create RELFILEs & SMPMCS that have the latest code. At the moment though it does NOT tag the SMPMCS that it matches a certain PTF/APAR level. So if it is difficult for your coding to decide if it is just GA or if it has a PTF/APAR/USERMOD to pick up (do we even support PSWI with an APAR/USERMOD build?), then we can look into enhancing the RELFILE build (smpe-fmid.sh) to include PTF info, and then all you need is the RELFILEs for PSWI.

Hi Onno,
When is created SMP/E environment, I know there are used RELFILEs. But I do not understand how this could help us with PTFs?

With support of APAR/USERMOD in PSWI.
1)With current pipeline, after PTF2 is applied, you can stop the pipeline, and fire receive/apply/accept from JCL as you normally do in SMP/E environment, after APAR/USERMOD is applied, you could continue with pipeline to build PSWI with APAR in it.

2) Or could be simply replicated step with PTF apply for specific APAR/USERMOD in the pipeline code.

3)Or user can use original PSWI with PTF1&PTF2, and when there is APAR/USERMOD for this particular build because of some error, he can simply run receive/apply/accept with that APAR against his SMP/E which was created by PSWI

So this is not issue at all (Option 3 is already available to solve this without need of any code change or manual action).

**Comment by OnnoVdT** [comment]
For you PSWI, you currently install RELLFILE, and then apply PTF. During a build, the RELFILE build step has access to the PTF data, so it can be augmented to update the SMPMCS marking that the current build includes the current PTF. This means that when you install those RELFILEs, there is no need anymore to RECEIVE/APPLY/ACCEPT the PTF, as SMPE already knows the PTF is included. So PSWI creation will be simplified (only need the current RELFILEs).
The SMPE build engine was created in 2 stages, with the service (PTF) part coming last, so this support was not added to the RELFILE build from the start, but it can be useful.

**Comment by jackjia-ibm** [comment]
Sorry I can not agree on your `release-zowe, build pswi, test pswi, potentially patch release` view of process. The reason I say this does not fit in this pipeline is,

- all steps in this pipeline are related, the following steps (test, publish) are handling the zowe build (convenience build, FMID/PTF/APAR/SYSMOD, docker images) created specifically for this branch/PR.
- your build inputs/outputs are not related to previous steps but more static and rely on what's put on the build server.

And with your view of release flow, the step also shouldn't be in this pipeline, because after Zowe is released, with `Jenkinsfile.release`, we don't rebuild Zowe with this `Jenkinsfile` pipeline anymore, then your step has no way to execute properly.

For your concern this branch could be accidentally deleted, I added protection on this branch to avoid.

So here are the 2 choices:

1. If you think this flow works for you, then please remove this build step from the current Jenkinsfile and create a new one. We can help you setup a Jenkins job to run this build. Then you team needs to be fully responsible to build and test PSWI artifact. Tom can only help you to move or upload to Artifactory.
2. If you think the PSWI build should follow same build/test/release process like other Zowe artifacts we release, then this PR has to wait until those issues are resolved. These are the changes I think are required:
  a) the PSWI build should take the Zowe SMPE build (could be FMID/PTF/APAR/SYSMOD) generated in previous step as inputs,
  b) Optionally this build should be automatically handle no matter 1 or 2 PTFs generated by Zowe. There are chances Zowe only generate 1 PTF.
  c) Optionally this build should support parallel building, otherwise the build server should be locked to avoid 2 PSWI building at same time
  d) The build output should be uploaded to Artifactory for future automated testing
  e) In the future, an automated test step should be added. The test should take the PSWI build generated in current pipeline, deploy and verify.  but not necessary be in this PR.
  f) In the future, the PSWI build should be added to release/promote pipeline so it can be promoted along with other Zowe build formats.

The change request is not about complaining your team don't have the knowledge to solve some of the problems, but just try to make the PSWI build process fits in our existing process. And I actually worked on a PR to solve 2.d step.

**Comment by muzikovam** [comment]
I will try to go with the second option, but it will take me some time. I will also have a lot of questions 🙂

**Pull request closed without merging**